scripts/lib/downloadImages.ts

```diff
@@ -17,23 +17,37 @@ import { createWriteStream } from "node:fs";
 import { finished } from "stream/promises";
 import { Readable } from "stream";
 import pMap from "p-map";
+import { startWebServer, closeWebServer } from "../lib/webServer";
 
 export async function downloadImages(
   images: Array<{ src: string; dest: string }>,
+  imagesPath: string,
 ) {
-  await pMap(
-    images,
-    async (img) => {
-      if (await pathExists(img.dest)) return;
-      const response = await fetch(img.src);
-      if (response.ok) {
-        await mkdirp(dirname(img.dest));
-        const stream = createWriteStream(img.dest);
-        await finished(Readable.fromWeb(response.body as any).pipe(stream));
-      } else {
-        console.log(`Error downloading ${img.src} to ${img.dest}`);
-      }
-    },
-    { concurrency: 4 },
+  const missingImages = images.filter(
+    async (img) => !(await pathExists(img.dest)),
   );
```

Comment:
Unlike the snippet I sent you on Slack, this will not run things in parallel. But it's probably fast enough to not matter.

--

Either way, this works properly?

Reply:
Oops, it's not working because the promise is always true and we always download the images. Good catch! With the snippet you sent me on Slack we have the problem that the array becomes of the type `img | null`, and I tried to do it differently. However, it seems it's not possible to use `filter` directly, thus I used your code but type casting the argument to the type we want.